### ISSUE TITLE
Needed: scalable way of selecting data from object stores for import into history

### ISSUE THREAD

**Issue by wm75** [galaxyproject/galaxy#11523]
I'm opening this issue to start a discussion how this could be implemented.

There are lots of different ways to get data into Galaxy already, but, as with so many things, the vast amounts of recently generated SARS-CoV-2 sequencing data have posed new challenges, and IMO revealed a gap in upload support:
while it's possible to upload data from various object stores as either individual datasets or as a collection, it is *not* possible  to filter the data in a scalable way before uploading it.

Things that *do* work currently:
- manually select datasets in the Upload manager under *Choose remote files*
  (doesn't scale beyond a handful of datasets)
- do the same in the Collections tab of the Upload manager
  (doesn't scale beyond a handful of datasets)
- prepare a list of URLs of resources you want and use them with the rule-based uploader
  (only works if you know URIs to your data)
- manually select datasets in the DL and use rule-based upload
  (doesn't scale beyond a handful of datasets)
- upload everything in a DL folder using rule-based upload, *then* filter the collection against a list of element identifiers (highly inefficient for large collections of data)

Some initial ideas that @bgruening and me have discussed a bit:
- extend remote files support of the rule-based uploader
  currently it presents URIs only of the top-level directory in an object store, but this should become navigable
- within the rule-based uploader provide support for filtering against the contents of a history dataset (implemented as an additional Filter action that brings up a column selector and a history dataset select box)
- an alternative (probably less elegant) would be to offer a way to upload URIs corresponding to remote files. Then these links could be filtered with standard Galaxy tools and passed back to the rule-based uploader.
- ideally, the rule-based uploader functionality would also be offered as a tool to become usable in WFs.

- another, more fancy idea was that you could be able to provide a list of checksums of datasets that should be uploaded and Galaxy would go and locate these files wherever they are. Question is, of course, what that checksum or unique ID should be since it would have to be unique but also known to users that want to obtain the data.

Anyway, what do you all think about this?

**Comment by jmchilton** [comment]
> extend remote files support of the rule-based uploader
currently it presents URIs only of the top-level directory in an object store, but this should become navigable

There is a little arrow key to descend into a directory - so it is not limited to the top-level directory:

<img width="1081" alt="Screen Shot 2021-03-02 at 1 03 54 PM" src="https://user-images.githubusercontent.com/216771/109693599-d9b7ba80-7b57-11eb-90d1-c509f351fc4b.png">

Does the UI need to be improved here or am I misunderstanding?

The backend supports recursively grabbing all the files below a directory - that should be a checkbox in here as well. I will implement that.

> within the rule-based uploader provide support for filtering against the contents of a history dataset (implemented as an additional Filter action that brings up a column selector and a history dataset select box)

This works right? Put URIs in a dataset and select them? Can you clarify this option?

<img width="217" alt="Screen Shot 2021-03-02 at 1 06 47 PM" src="https://user-images.githubusercontent.com/216771/109693908-30bd8f80-7b58-11eb-84c1-c2c33dacd1de.png">

Do you mean filtering against a whole history instead of a history dataset?

You can filter datasets with search and then select them in the rule builder. The problem here is pagination if more than 500 datasets need to be selected. I think that is being worked on as part of the history rewrite but we could also address it by adding an option when selecting datasets in the rules.

> an alternative (probably less elegant) would be to offer a way to upload URIs corresponding to remote files. Then these links could be filtered with standard Galaxy tools and passed back to the rule-based uploader.

I mean again I think this works. If it doesn't there should be an issue and we should just fix it.

> ideally, the rule-based uploader functionality would also be offered as a tool to become usable in WFs.

Yes, I agree I think. Seems we have a lot of the pieces to make this work.

> another, more fancy idea was that you could be able to provide a list of checksums of datasets that should be uploaded and Galaxy would go and locate these files wherever they are. Question is, of course, what that checksum or unique ID should be since it would have to be unique but also known to users that want to obtain the data.

It is a fun idea but maybe its own issue? It is a lot harder and farther a field from the rest of these?

**Comment by wm75** [comment]
> > extend remote files support of the rule-based uploader
> > currently it presents URIs only of the top-level directory in an object store, but this should become navigable
>
> There is a little arrow key to descend into a directory - so it is not limited to the top-level directory:

Ok, top-level directory wasn't a good term here. What I meant was only one level deep. As soon as you navigate into any of the folders in your screenshot only the URIs of the files therein will be displayed.

**Comment by wm75** [comment]
> > within the rule-based uploader provide support for filtering against the contents of a history dataset (implemented as an additional Filter action that brings up a column selector and a history dataset select box)
>
> This works right? Put URIs in a dataset and select them? Can you clarify this option?

Yes, you can use the URIs in a tabular dataset, then construct additional columns from those URIs and build a nice collection.
What I meant though is that after having generated columns to be used as the URI, the list identifier and the PE indicator, and possibly more, you may want to use a *second* file the contents of which would be matched against any of these columns, and only rows with (or without match would be kept (kind of a Filter list from contents of a file within the rule-builder).

**Comment by wm75** [comment]
You *can* of course do this in two-steps already now, by first using the rule-based uploader to build a huge collection, then use the Filter list tool to keep only the elements you're interested in, but for really huge collections of data you'd really prefer to have the filtering done *before* you retrieve the data.

**Comment by bgruening** [comment]
@jmchilton maybe you have a better idea for our is our use-case:

We have 200.000 files and downloading them is error-prone. So we download/rsync them to our server and would like to get it into Galaxy. A subset of those files is now interesting, 22.000 files (paired) for now, and we would like to get it into history with collections. How can we subset the 200.000 files and just take a few out of that?

What we tried today is to copy those 22.000 files into Wolfgangs FTP dir and use the rule-builder to create a collection. After 12h this crashed, no clue why. I see in the job_working_dir 22.000 files but that's it.

**Comment by wm75** [comment]
> > extend remote files support of the rule-based uploader
> > currently it presents URIs only of the top-level directory in an object store, but this should become navigable
> 
> There is a little arrow key to descend into a directory - so it is not limited to the top-level directory:
> 
> <img alt="Screen Shot 2021-03-02 at 1 03 54 PM" width="1081" src="https://user-images.githubusercontent.com/216771/109693599-d9b7ba80-7b57-11eb-90d1-c509f351fc4b.png">
> 
> Does the UI need to be improved here or am I misunderstanding?
> 

@jmchilton This works exactly as you are describing it now! Maybe this fixed itself with the update of usegalaxy.eu to 21.01.

As an additional suggestion, a preview of the file contents of a directory would be helpful.
Right now you have to sort of blindly click on a folder to select its files and hope it contains whaat you are looking for.
If it doesn't you have to repeat all the navigation through folders.
